v2/cf-nodejs-logging-support/src/lib/logger/cache-factory.ts

```diff
@@ -0,0 +1,66 @@
+import Config from "../config/config";
+import { ConfigField, outputs } from "../config/interfaces";
+import { REDACTED_PLACEHOLDER } from "./record-factory";
+import { SourceUtils } from "./source-utils";
+
+export default class CacheFactory {
+
+    private static instance: CacheFactory;
+    private config: Config;
+    private sourceUtils: SourceUtils;
+
+    private constructor() {
+        this.config = Config.getInstance();
+        this.sourceUtils = SourceUtils.getInstance();
+    }
+
+    public static getInstance(): CacheFactory {
+        if (!CacheFactory.instance) {
+            CacheFactory.instance = new CacheFactory();
+        }
+
+        return CacheFactory.instance;
+    }
+
+    createCache(output: outputs, req?: any, res?: any): any {
+        const writtenAt = new Date();
+
+        let cachedFields: ConfigField[];
+        if (output == "msg-log") {
+            cachedFields = this.config.getCacheMsgFields();
+        } else {
+            cachedFields = this.config.getCacheReqFields();
+        }
+
+        let cache: any = {};
+        cachedFields.forEach(
+            field => {
+                if (!Array.isArray(field.source)) {
+                    if (output == "msg-log") {
+                        cache[field.name] = this.sourceUtils.getFieldValue(field.source, cache, writtenAt);
+                    } else {
+                        cache[field.name] = this.sourceUtils.getReqFieldValue(field.source, cache, writtenAt, req, res);
+                    }
+                } else {
+                    const value = this.sourceUtils.getValueFromSources(field, cache, output, writtenAt, req, res);
+
+                    if (value != null) {
+                        cache[field.name] = value;
+                    }
+                }
```

Comment:
It makes sense to keep the cache in its own file/class. But providing data to the cache as part of the CacheFactory code does not really follow the principle of separation of concerns IMO. We can discuss this.

Reply:
I applied the changes as discussed.